sandbox/cgroup/freezer.go

```diff
@@ -100,6 +112,75 @@ func thawSnapProcessesImpl(snapName string) error {
 	return nil
 }
 
+func applyToSnap(snapName string, action func(groupName string) error) error {
+	canary := fmt.Sprintf("snap.%s.", snapName)
+	cgroupRoot := filepath.Join(rootPath, cgroupMountPoint)
+	if _, dir, _ := osutil.DirExists(cgroupRoot); !dir {
+		return nil
+	}
+	return filepath.Walk(filepath.Join(rootPath, cgroupMountPoint), func(name string, info os.FileInfo, err error) error {
+		if !info.IsDir() {
+			return nil
+		}
+		if !strings.HasPrefix(info.Name(), canary) {
+			return nil
+		}
+		// found a group
+		if err := action(name); err != nil {
+			return err
+		}
+		return filepath.SkipDir
+	})
+}
+
+// freezeSnapProcessesImplV1 freezes all the processes originating from the given snap.
+// Processes are frozen regardless of which particular snap application they
+// originate from.
+func freezeSnapProcessesImplV2(snapName string) error {
+	freezeOne := func(dir string) error {
+		fname := filepath.Join(dir, "cgroup.freeze")
+		if err := ioutil.WriteFile(fname, []byte("1"), 0644); err != nil && os.IsNotExist(err) {
+			//  the group may be gone already
+			return nil
+		} else if err != nil {
+			return fmt.Errorf("cannot freeze processes of snap %q, %v", snapName, err)
+		}
+		for i := 0; i < 30; i++ {
+			data, err := ioutil.ReadFile(fname)
+			if err != nil {
+				return fmt.Errorf("cannot determine the freeze state of processes of snap %q, %v", snapName, err)
+			}
+			// If the cgroup is still freezing then wait a moment and try again.
+			if !bytes.Equal(bytes.TrimSpace(data), []byte("1")) {
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+		}
+		return nil
```

Comment:
If we got here, it might also be because the 3 seconds timeout has passed and we have given up re-trying. Shouldn't we return an error in that case?

Reply:
refactored ths bit